bpf/include/bpf/builtins.h

```diff
@@ -10,12 +10,80 @@
 # define lock_xadd(P, V)	((void) __sync_fetch_and_add((P), (V)))
 #endif
 
+#define __it(x, op) (x -= sizeof(__u##op))
+
+static __always_inline void __bpf_ua_memset(void *d, const __u8 c,
+					    const __u64 len)
+{
+	switch (len) {
+#define __it_set(a, b, op) (*(__u##op *)__it(a, op)) = b
```

Comment:
Don't we need to copy `b` several times since it's `__u8` and we want to write a `__u{64,32}`?

Reply:
I haven't looked at the output of `__bpf_ua_memset()` yet, only the `__bpf_ua_memcpy()` so far, so I'm going to do that now. :-)

Reply:
The output looks reasonable to me, for example, clang generates:
```
       1:	b7 04 00 00 00 00 00 00	r4 = 0
      [...]
      15:	7b 43 f8 ff 00 00 00 00	*(u64 *)(r3 - 8) = r4
      16:	7b 43 f0 ff 00 00 00 00	*(u64 *)(r3 - 16) = r4
      17:	7b 43 e8 ff 00 00 00 00	*(u64 *)(r3 - 24) = r4
      18:	7b 43 e0 ff 00 00 00 00	*(u64 *)(r3 - 32) = r4
      19:	7b 43 d8 ff 00 00 00 00	*(u64 *)(r3 - 40) = r4
      20:	7b 43 d0 ff 00 00 00 00	*(u64 *)(r3 - 48) = r4
      21:	7b 43 c8 ff 00 00 00 00	*(u64 *)(r3 - 56) = r4
      22:	7b 43 c0 ff 00 00 00 00	*(u64 *)(r3 - 64) = r4

```

Reply:
Because `r4 = 0`. If you pass a value, you'll end up writing only 8 bits of those 64; the others will be remain zero. Or am I missing something?

Reply:
I'm not sure I understand your concern yet. We're doing a `*(u64 *)(r3 - 8) = ...`, why would we end up only assigning 1 byte?

Reply:
I talked about it with Paul and there are two points that we didn't manage to figure out.

1) As Paul mentioned, the constant copied is a `__u8` and it seems that would result in a value up to 255 being assigned to a 64-bit integer. With your example:
    ```
       1:	b7 04 00 00 00 00 00 00	r4 = 3    // changed here
      [...]
      15:	7b 43 f8 ff 00 00 00 00	*(u64 *)(r3 - 8) = r4
      16:	7b 43 f0 ff 00 00 00 00	*(u64 *)(r3 - 16) = r4
      17:	7b 43 e8 ff 00 00 00 00	*(u64 *)(r3 - 24) = r4
    ```
   It feels like we will end up with
    ```
    r3 == 00 00 00 00 00 00 00 03   00 00 00 00 00 00 00 03   00 00 00 00 00 00 00 03
    ````
    Instead of setting each byte:
    ```
    r3 == 03 03 03 03 03 03 03 03   03 03 03 03 03 03 03 03   03 03 03 03 03 03 03 03
    ````

2) We also struggle to understand the subtraction in `#define __it(x, op) (x -= sizeof(__u##op))`. It makes this new `memset()` start from the beginning of the destination and go backwards. Should it not go forwards with a `+` instead?

Reply:
Re 1) Ahh, good point. I see now what you mean. Yep, I'll change the name to bzero() or so, since we don't need anything else other than zero'ing in our code.

Re 2) Yep, I just noticed here. Missing an initial ptr+=len, since we walk backwards.

Reply:
> since we don't need anything else other than zero'ing in our code.

Except for [`test_ipv6_addr_clear_suffix()`](https://github.com/cilium/cilium/blob/f96afb0821ca95a23429704604e8707c043ec94d/bpf/lib/ipv6_test.h#L9), the failing Travis CI test, but we probably don't really **need** it.

Reply:
Fyi, still figuring out what we can do wrt stack alignment. Sadly the verifier forces BPF access to stack as aligned. Digging up a bit the LLVM code to see if there's a reasonable workaround.